### PR TITLE
Implemented Sprague Grundy Theorem on Strings

### PR DESCRIPTION
**Pull request by KartikeySharma**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

Sprague-Grundy Theorem says that in a composite game in which 2 players A and B play, if both A and B play optimally (i.e., they don’t make any mistakes), then the player starting first is guaranteed to win if the XOR of the grundy numbers of position in each sub-games at the beginning of the game is non-zero. Otherwise, if the XOR evaluates to zero, then player A will lose definitely, no matter what.

### Checklist

- [Tick] I've read the contribution guidelines.
- [Tick] I've checked the issue list before deciding what to submit.
- [Tick] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Issue #1701

### CODE REVIEW

**Comment by KartikeySharma** [comment]
Pls review @atarax665

**Comment by KartikeySharma** [comment]
Pls check @atarax665

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    cin>>a>>y;
+    for(g=0;g<y;g++){
+        cin>>wq[g];
+        process(g);
+    }
+    // Run the Sprague Grundy algo on each recursive combination
+    re=sprague_grundy(0,a.size()-1);
+    // Outputting which player wins
+    if(re==1)
+    	cout<<"Player 1";
```
Comment (atarax665) [inline C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp]:
```suggestion
    	cout<<"Player 1 wins.";

```

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    else
+    	cout<<"Player 2";
```
Comment (atarax665) [inline C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp]:
```suggestion
    	cout<<"Player 2 wins";

```

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+int main()
+{    
+    int g,h,re;
+    // Intitialising the array with value of INF
+    for(g=0;g<=30;g++) {
+        for(h=0;h<=30;h++) {
+            kk[g][h]=INF;
+        }
+    }
+    cin>>a>>y;
```
Comment (atarax665) [inline C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp]: Prompt user to input the required data.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Format your code according to C++ style guide. You may use [this](https://techiedelight.com/tools/clike) for it.

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp (hunk 1 above)
Comment (KartikeySharma) [reply C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp]: ok

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp (hunk 3 above)
Comment (KartikeySharma) [reply C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp]: ok

**Comment by KartikeySharma** [comment]
Pls review @atarax665

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	else
+		cout << "Player 2 wins.";
+
+	return 0;
+}
+
+/*
+Time Complexity: O(n^2)
+Sample Input:
+spraguegrundy	// Main string
```
Comment (atarax665) [inline C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp]: Update the Sample Input/Output according to the updated code.

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
+/*The game used to implement Sprague-Grundy is defined below
+
+ -> There are 2 people playing this composite impartial game
+ -> The second player writes an arbitrary string . 
+ -> Then each player starting with the first player starts 
+ 	removing a continous substring which exists in a list of 
+ 	substrings that are possible to remove.
+ -> The first player who can't remove a substring loses. */
```
Comment (atarax665) [inline C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp:8]: Add a line break after this line.

**Review by atarax665 — CHANGES_REQUESTED** [review]

**Comment by KartikeySharma** [comment]
@atarax665 Pls review. I have other PRs to submit too. Pls, understand as it has been 2 days since you last replied.

**Comment by atarax665** [comment]
> @atarax665 Pls review. I have other PRs to submit too. Pls, understand as it has been 2 days since you last replied.

You haven't resolved the last suggested changes yet. Please do the suggested changes ASAP and request a review.

**Comment by KartikeySharma** [comment]
Pls have a look now @atarax665

File: C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		cout << "Player 1 wins.";
+	else
+		cout << "Player 2 wins.";
+
+	return 0;
+}
+
+/*
+Time Complexity: O(n^2)
+Sample Input:
```
Comment (atarax665) [inline C-Plus-Plus/dp/Sprague_Grundy_Theorem.cpp:142]: Sample Input/Output should be exactly what shows on the screen when the program is run.
output
Format it this way.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Also resolve the merge conflict so that I can merge this PR.

**Review by atarax665 — APPROVED** [review]
I have made the appropriate changes as it was just a minor one, please make sure to resolve such minor issues ASAP so that one of the maintainer can review and merge the PR.

**Comment by atarax665** [comment]
Resolve the conflict ASAP so that I can merge this PR.

**Comment by KartikeySharma** [comment]
I don't know how this conflict arose. Could you please help me in understanding as I have never encountered such an error before...

**Comment by KartikeySharma** [comment]
Resolved the issue @atarax665 . Pls merge it . Thanks :)

**Comment by atarax665** [comment]
:+1:

**Comment by KartikeySharma** [comment]
Pls add DWOC tag @atarax665 .. i forgot to mention it earlier

**Pull request merged**